core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -176,6 +174,50 @@ public static StreamProperties deriveProperties(
         return result;
     }
 
+    /**
+     * Determines whether a local exchange is single-stream distributed at its only input source. This method will is expensive
+     * since it requires traversing the entire sub-plan at each local exchange node, so calling this method should be avoided
+     * whenever possible and new usages should not be added.
+     *
+     * @param exchangeNode a local exchange with a single input source to check for single stream input distribution
+     * @throws IllegalArgumentException if the exchange is not a local exchange or does not have only a single input source
+     * @deprecated Only for use by {@link PropertyDerivations}
+     */
+    @Deprecated
+    static boolean isLocalExchangesSourceSingleStreamDistributed(ExchangeNode exchangeNode, Metadata metadata, Session session)
+    {
+        checkArgument(exchangeNode.getScope() == LOCAL, "exchangeNode must be a local exchange");
+        checkArgument(exchangeNode.getSources().size() == 1, "exchangeNode must have a single source");
+
+        return deriveStreamPropertiesWithoutOtherActualPropertiesRecursively(exchangeNode.getSources().get(0), metadata, session).isSingleStream();
+    }
+
+    /**
+     * Derives {@link StreamProperties} without populating {@link StreamProperties#otherActualProperties}. This is necessary to avoid exponential-time,
+     * mutually recursive sub-plan traversals when {@link PropertyDerivations} attempts to check a local exchange's input source for single stream distribution.
+     *
+     * @deprecated For internal use only by {@link StreamPropertyDerivations#isLocalExchangesSourceSingleStreamDistributed(ExchangeNode, Metadata, Session)}
+     */
+    @Deprecated
+    private static StreamProperties deriveStreamPropertiesWithoutOtherActualPropertiesRecursively(PlanNode node, Metadata metadata, Session session)
+    {
+        List<StreamProperties> inputProperties = node.getSources().stream()
+                .map(source -> deriveStreamPropertiesWithoutOtherActualPropertiesRecursively(source, metadata, session))
+                .collect(toImmutableList());
+
+        return deriveStreamPropertiesWithoutOtherActualProperties(node, inputProperties, metadata, session);
+    }
+
+    private static StreamProperties deriveStreamPropertiesWithoutOtherActualProperties(PlanNode node, List<StreamProperties> inputProperties, Metadata metadata, Session session)
```

Comment:
nit: `rnn` on args?

Reply:
I think we can omit the checks since this is an internal private method called with relatively high frequency